cpp/inc/bond/core/detail/check_overflow.h

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include <limits>
+#include <stdexcept>
+
+namespace bond
+{
+namespace detail
+{
+
+    inline void check_add_overflow(const char* ptr, uint32_t offset)
+    {
+        if ((ptr + offset) < ptr)
```

Comment:
I'm not sure whether this arithmetic can trigger UB or not... Let me dig into this some more.